torch/distributed/_shard/sharded_tensor/api.py

```diff
@@ -273,46 +278,85 @@ def gather(
                 Must to be provided ONLY on ``dst`` rank.
                 Default: ``None``
         """
+        def shard_size(shard_md):
+            res = 1
+            for s in shard_md.shard_sizes:
+                res *= s
+            return res
         rank = dist.get_rank(self._process_group)
         full_size = self.metadata().size
         _validate_output_tensor_for_gather(rank, dst, full_size, out)
 
         local_shards = self.local_shards()
 
         world_size = dist.get_world_size(self._process_group)
-
-        gathered_shards: List[Optional[List[Shard]]] = [None] * world_size if rank == dst else []
-        # TODO: see how we could use dist.gather() instead of dist.gather_object
-        # as the latter one involves pickling on CPU, see more context
-        # https://github.com/pytorch/pytorch/issues/73935
-        dist.gather_object(
-            obj=local_shards,
-            object_gather_list=gathered_shards,
+        rank_sizes = [0 for _ in range(world_size)]
+        max_rank_size = 0
+        shard_placement: Dict[int, Tuple[int, int]] = dict()
+        local_shards_placement = []
+        # collect sizes
+        for shard_idx, shard_md in enumerate(self.metadata().shards_metadata):
+            shard_rank = cast(_remote_device, shard_md.placement).rank()
+            assert shard_rank is not None
+            shard_placement[shard_idx] = (shard_rank, rank_sizes[shard_rank])
+            if shard_rank == rank:
+                local_shards_placement.append((shard_md, rank_sizes[shard_rank],))
+
+            rank_sizes[shard_rank] += shard_size(shard_md)
+            max_rank_size = max(max_rank_size, rank_sizes[shard_rank])
+
+        gather_list: Optional[List[torch.Tensor]]
+        if rank == dst:
+            assert out is not None
+            gather_list = [torch.empty((max_rank_size,), device=out.device) for _ in range(world_size)]
+        else:
+            gather_list = None
+
+        with torch.no_grad():
+            # This is an aproximation for what device to use.
+            # The lack of policy enforcement around is pretty bad
+            device: Union[torch.device, str]
+            if len(local_shards) > 0:
+                device = local_shards[0].tensor.device
+            else:
+                device = f"cuda:{torch.cuda.current_device()}"
```

Comment:
Why don't we always use the current device? We do expect the user to set this appropriately as documented here: https://github.com/pytorch/pytorch/blob/master/torch/distributed/_shard/sharded_tensor/api.py#L109

Reply:
Always using CUDA is wrong if this is a CPU ShardedTensor.

What we want is to alloc on `self._process_group` device which should be `cuda.current_device` if we're using NCCL or cpu with GLOO.

What I'm thinking here, and I need to add a test for, is how to handle `sharded_tensor.cpu().gather()` when there's one rank without data.

Reply:
I replaced this with adding a method `_get_preferred_device` to ST which returns cuda:current_device for NCCL or CPU otherwise.